runtime/compiler/wasm/leb128_test.go

```diff
@@ -0,0 +1,168 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright 2019-2020 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package wasm
+
+import (
+	"math"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestBuf_writeULEB128(t *testing.T) {
+
+	t.Run("DWARF spec", func(t *testing.T) {
+
+		// DWARF Debugging Information Format, Version 3, page 140
+
+		for v, expected := range map[uint32][]byte{
+			2:     {2},
+			127:   {127},
+			128:   {0 + 0x80, 1},
+			129:   {1 + 0x80, 1},
+			130:   {2 + 0x80, 1},
+			12857: {57 + 0x80, 100},
+		} {
+			var b buf
+			err := b.writeULEB128(v)
+			require.NoError(t, err)
+			require.Equal(t, expected, b.data)
+
+			b.offset = 0
+
+			actual, err := b.readULEB128()
+			require.NoError(t, err)
+			require.Equal(t, v, actual)
+		}
+	})
+
+	t.Run("write: max byte count", func(t *testing.T) {
+		var b buf
+		err := b.writeULEB128(math.MaxUint32)
+		require.NoError(t, err)
+		require.GreaterOrEqual(t, max32bitLEB128ByteCount, len(b.data))
+	})
+
+	t.Run("read: max byte count", func(t *testing.T) {
+		b := buf{data: []byte{0x81, 0x82, 0x83, 0x84, 0x85, 0x86, 0x87, 0x88}}
```

Comment:
Do these bytes represent the biggest value can be read?
isn't biggest uint32 can be represented as `0xFF, 0xFF, 0xFF, 0xFF`?

Reply:
No they don't, note it says "max byte count" not "max byte value". This test ensures that only up to the maximum number of bytes are read when reading a LEB128-encoded 32-bit number, which is 5 (see https://github.com/onflow/cadence/blob/ea6b35ba72b9622a814c3d27d718d120a594f06f/runtime/compiler/wasm/leb128.go#L145-L151), i.e. the function checks that only 5 out of the 8 bytes are read, to ensure the LEB128 parser doesn't keep reading infinitely.

I've added comments to clarify this in 4e7e058c9fd16e4884825d3db3cf9b768b70eb3a